README.md
# Reddit-Shopping-Tool
Reddit Shopping Tool is an extension that automatically identifies any users that have been banned on the <a href="https://www.reddit.com/r/UniversalScammerList/wiki/banlist">Universal Scammer List</a> 

<a href="https://chrome.google.com/webstore/detail/reddit-shopping-tool/pimdepbkfokgeadmhmhfpapfdbodadlg">Chrome Webstore</a>
<br>
<a href="https://addons.mozilla.org/en-US/firefox/addon/reddit-shopping-tool/">Firefox AMO Store</a>

<strong>Big disclaimer: this does not identify any new scammers or scammers that have reddit accounts not on the list.</strong>
<hr>

# Setup
1. Clone/download git repo
2. Open Chrome extension menu
3. Enable developer mode
4. Click "Load unpacked" and select the repo folder

<strong>Change Log</strong>

Version 1.5.4 / 2020-9-12
* Replaced source links to comply with CORS again

Version 1.5.3 / 2019-4-18
* Replaced source links to comply with CORS
* Fixed bug with updating the list
* Removed tooltip/info box, and replaced with link to universalscammerlist.com
* Added support for old and new reddit
* General bug fixes

Version 1.5.2 / 2019-4-2
* Fixed source links
* Replaced r/UniversalScammerList source with r/RSTList
* Clean up from last update

Version 1.5.1 / 2019-3-18
* Fixed bug with manifest/permission
* Removed subreddit restrict option
* Changed ban list update intervals
* General bug fixes

Version 1.5.0 / 2019-1-2
* Created source https://www.reddit.com/r/RSTList/wiki/banlist which contains universalscammerlist.com data 
* Checking the new source
* Removed options for hourly, and 6 hours in update frequency
* General improvements

changelog.txt
Version 1.5.4 / 2020-9-12
* Replaced source links to comply with CORS again

Version 1.5.3 / 2019-4-18
* Replaced source links to comply with CORS
* Fixed bug with updating the list
* Removed tooltip/info box, and replaced with link to universalscammerlist.com
* Added support for old and new reddit
* General bug fixes

Version 1.5.2 / 2019-4-2
* Fixed source links
* Replaced r/UniversalScammerList source with r/RSTList
* Clean up from last update

Version 1.5.1 / 2019-3-18
* Fixed bug with manifest/permission
* Removed subreddit restrict option
* Changed ban list update intervals
* General bug fixes

Version 1.5.0 / 2019-1-2
* Created source https://www.reddit.com/r/RSTList/wiki/banlist which contains universalscammerlist.com data 
* Checking the new source
* Removed options for hourly, and 6 hours in update frequency
* General improvements

Version 1.4.0 / 2018-1-19
* Added support for new profile page
* Checking new scammer list at https://www.reddit.com/r/hardwareswap/wiki/banlist
* Added option to check for sketchy users (disabled by default)
* General improvements and bug fixes

Version 1.3.2 / 2017-7-7
* Fixed issues with dependencies
* Switched to local storage for Firefox support
* Made changes to fix unsanitized data
* Changed name to "Reddit Scam Labeler", because the other title was misleading (according to Chrome)

Version 1.3.1 / 2017-2-11
* Fixed bug where extension was not running in inbox

Version 1.3.0 / 2017-1-28
* Fixed typos on the subreddit list
* Added badge indicators
* Added more info box when clicking on the badge
* Added parsing for subreddit and reason data
* Added last update time at the data tab in the options page
* Added Save Settings button to top of options page
* Added option to disable more info box
* Added contact link
* Added support for Firefox
* Minor bug fixes

Version 1.2.1 / 2016-8-1
* Fixed bug where list was not updating

Version 1.2.0 / 2016-7-10
* Added subreddit filter
* Added option to edit filter list
* Fixed: update list when extension updates

Version 1.1.0 / 2016-7-1
* New tags [SKETCHY], [TROLL], [ACCOUNT COMPROMISED]
* Added RES never ending support (disabled by default, enable it in options)
* Added Options page
* Added option to change ban list update time
* Added option to enable and disable RES
* Added force update button
* Added data tab in options to view current ban list
* Added info tab
* Minor bug fixes

Version 1.0.1 / 2016-6-13
* Minor bug fixes

Version 1.0.0 / 2016-6-13
* Initial release

setoptions.js
var defaultoptions = 
{
	"updateTime": 360, 
	"labelSketchy":0
};
//to ensure it only loads once
var loaded = false;
$(document).ready(function(){
	//load saved options, or use defaults
	 chrome.storage.local.get({ 
		options: defaultoptions
	 }, function(data) {
		 options = data.options;
		 for (var optionname in data.options){
			if (data.options.hasOwnProperty(optionname)) {
				//set defaults
				$("#" + optionname).val(data.options[optionname]).change();
				//initialize delete button, has to be in callback instead of document.ready
				$(".delete-btn").click(function(){
					$(this).parent().remove();
				});
			}
		}
	});
	initBtn();
	$(".version").html(version);
	$('.changelog').load("changelog.txt");
});
function loadUsers(){
	checkForUpdate();
	chrome.storage.local.get(['users','timestamp'], function(data){
		//loop through banned users
		users = data.users;
		var date = (Date.now() - data.timestamp)/1000; //seconds
		var ago = (Math.floor(date/86400) > 0) ? Math.floor(date/86400) + " days ago" : Math.round(date/3600) + " hours ago";
		$("#lastUpdate").html("Last Updated: " + ago);
		for (var name in users){
			if (users.hasOwnProperty(name)) {
				//add to textarea display
				$(".banned-users").append(document.createTextNode(name + " - " + users[name] + "\r\n"));
				
			}
		}
	});
}
//Initialize all buttons
function initBtn(){
	$("#data-tab").click(function(){
		//load the banned user data list when requested
		if(!loaded){
			loadUsers();
			loaded = true;
		}
	});
	$("#forceupdate").click(function(){
		$(this).text("Updating..");
		$(this).addClass("disabled");
		updateList(function(msg){
			alert(msg);
			$("#forceupdate").text("Force Ban List Update");
			$("#forceupdate").removeClass("disabled");
		});
	});
	$("#setdefault").click(function(){
		chrome.storage.local.set({
			options: defaultoptions
		}, function() {
			//display success message
			var status = $("#status");
			status.html('<div class="alert alert-success">Options Successfully Saved</div>');
			setTimeout(function() {
			  status.html('');
			}, 5000);
		});
		window.location.reload();
	});
	
	$("#save").click(function() {
		var optionset = {};
		//loop through for any options
		$(".rst-option").each(function() {
			//add to options array
			var value = $(this).val();
			var optionname = $(this).attr('id');
			optionset[optionname] = value;
		});
		//save array to chrome sync'd storage
		chrome.storage.local.set({
			options: optionset
		}, function() {
			//display success message and scroll to top
			var status = $("#status");
			status.html('<div class="alert alert-success">Options Successfully Saved</div>');
			$(window).scrollTop(0);
			setTimeout(function() {
			  status.html('');
			}, 5000);
		});
		option = optionset;
	});
}

rst-script.js
var uservis = {};
var options = {};
var version = "1.5.3";
var defaultoptions = 
{
	"updateTime": 360, 
	"labelSketchy":0
};
var lastInfoBox;
//load options before running rest of script
$( document ).ready(()=>{
	chrome.storage.local.get({
		options: defaultoptions
	 }, function(data) {
		options = data.options;
		if (window.location.hostname.indexOf("reddit.com") != -1 ){
			checkForUpdate();
			labelUsers();
			checkForChanges();
			
		}
	});
});
function checkForUpdate(){
	//check last update timestamp
	chrome.storage.local.get(['timestamp','version'], function(data){
		//update list if list is X days old or empty or extension was updated
		if(data.timestamp === null || data.version === null || data.version !== version ||
		( options["updateTime"] !== -1 && ( ( Date.now() - data.timestamp )/1000 ) > ( options.updateTime * 60 ) ) ) {
			updateList();
		}
	});
}
function labelUsers(){
	//create an array for users that need to be checked
	$( ".author, .Post__username, .Comment__author, .Post__authorLink, a[href^='/user/']" ).each(function() {
		if(uservis[$(this).text().toLowerCase()] == null){
			uservis[$(this).text().toLowerCase()] = "";
		}
	});
	chrome.storage.local.get('users', function(data){
		//loop through banned users and check if uservis contains any banned users
		var dbUsers = data.users;
		for (var name in uservis){
			//set as banned if not already on the list
			if (uservis[name.toLowerCase()] <= 0 && dbUsers.hasOwnProperty(name)) {
				//set user info in uservis, and label sketchy users if enabled
				if(options['labelSketchy'] == 1 || dbUsers[name].code != 2){
					uservis[name.toLowerCase()] = dbUsers[name];
				}
			}
		}
		//automoderator exception
		uservis["automoderator"] = "";
		
		//loop through all name tags and set them as banned/sketchy, if any
		$( ".author, .Post__username, .Comment__author, .Post__authorLink, a[href^='/user/']" ).each(function() {
			
			//check if user is on the list
			var userData = uservis[$(this).text().toLowerCase()];
			if(userData){
				//translate bancode and add badge next to their name
				var badge = $("<a></a>");
				badge.addClass("rst-banned-" + userData);
				badge.addClass("rst-badge");
				badge.text(userData);
				badge.attr('href',"https://universalscammerlist.com/search.php?username=" + $(this).text().replace("/u/",""));
				$(this).append(badge);
			}
		});
	});
}

var sources = [
	"https://www.reddit.com/r/hardwareswap/wiki/banlist.json",
	"https://www.reddit.com/r/RSTList/wiki/banlist.json",  
	"https://www.reddit.com/r/RSTList/wiki/sketchylist.json"
];
function updateList(callback){
	var users = {};
	Promise.all( sources.slice(0, options["labelSketchy"] == 1 ? 3 : 2).map(function( src, index ) {
		return $.get( src );
	})).then(function(res) {
		var msg = "Ban List Updated!";
		var error = false;
		$.each(res,function( i, data ){
			if(data)
				$.extend(users, getUsersFromJSON(data));
			else{
				msg = "Error, failed to get data";
				error = true;
			}
		});
		if(!error){
			chrome.storage.local.set({"users": users});
			chrome.storage.local.set({"timestamp": Date.now()});
			chrome.storage.local.set({"version": version});
		}
		// console.log("[RST] " + msg);
		if( !!callback ) {
			callback(msg);
		}
	});
}
function getUsersFromJSON(data){
	var rx = /\*\s+\/u\/([^\s#]+)(\s+|)(#[a-z0-9]+)?/;
	if( typeof data === 'undefined' || !data.data || !data.data.content_md ) {
		return {};
	}
	var result = data.data.content_md.match( new RegExp( rx, 'gi') ) || [];
	result = result.map(function(i) {
		return i.match( new RegExp( rx, 'i' ) );
	});
	var ret = {};
	$.each(result,function(i,v){
		if( typeof v[3] === 'undefined' ) {
			v[3] = '#sketchy';
		}
		v[3] = v[3].slice(1).toLowerCase().replace('sktechy','sketchy');
		if( v[3].indexOf('ske') === 0 ) {
			v[3] = 'sketchy';
		}
		if( v[3].indexOf('spa') === 0 ) {
			v[3] = 'spammer';
		}
		if( v[3].indexOf('sca') === 0 ) {
			v[3] = 'scammer';
		}
		if( ['scammer','troll','sketchy','bot','compromised','rule','slavelabour','investigating','impersonator','spammer'].indexOf( v[3] ) === -1 ) {
			v[3] = 'sketchy';
		}
		ret[v[1].toLowerCase()] = v[3];
	});
	return ret;
}
//Checks for changes in the page, in dynamically loaded content such as infinite scroll
var lastcount = 0;
function checkForChanges()
{
	var namecounts = $( ".author, .Post__username, .Comment__author, .Post__authorLink, a[href^='/user/']" ).length;
	if(lastcount != namecounts){
		lastcount = namecounts;
		labelUsers();
	}
	//recheck every two second
	setTimeout(checkForChanges, 2000);
}